lang/py/src/avro/io.py

```diff
@@ -478,6 +573,77 @@ def write_crc32(self, bytes):
     """
     self.write(STRUCT_CRC32.pack(crc32(bytes) & 0xffffffff));
 
+  def write_date_int(self, datum):
+    """
+    Encode python date object as int.
+    It stores the number of days from
+    the unix epoch, 1 January 1970 (ISO calendar).
+    """
+    delta_date = datum - datetime.date(1970, 1, 1)
+    if delta_date.days < 0:
+      raise schema.AvroException("Cannot encode date before 1 January 1970 (ISO calendar)")
+    self.write_int(delta_date.days)
+
+  def _validate_time_datum(self, value, time_type):
+    if (time_type in [constants.TIME_MILLIS] and
+        not (UNSIGNED_INT_MIN_VALUE <= value <= UNSIGNED_INT_MAX_VALUE)):
+      raise schema.AvroException("Cannot encode %s logicaltype schema with value outside (%d, %d)." % (
+        time_type,
+        UNSIGNED_INT_MIN_VALUE,
+        UNSIGNED_INT_MAX_VALUE)
+      )
+    elif (time_type in [constants.TIME_MICROS, constants.TIMESTAMP_MILLIS, constants.TIMESTAMP_MICROS] and
+        not (UNSIGNED_LONG_MIN_VALUE <= value <= UNSIGNED_LONG_MAX_VALUE)):
+      raise schema.AvroException("Cannot encode %s logicaltype schema with value outside (%d, %d)." % (
+        time_type,
+        UNSIGNED_LONG_MIN_VALUE,
+        UNSIGNED_LONG_MAX_VALUE)
+      )
+
+  def write_time_millis_int(self, datum):
+    """
+    Encode python time object as int.
+    It stores the number of milliseconds from midnight, 00:00:00.000
+    """
+    offset_milliseconds = (((datum.hour * 60) + datum.minute)*60 + datum.second)*1000 + datum.microsecond // 1000
+    self._validate_time_datum(offset_milliseconds, constants.TIME_MILLIS)
+    self.write_int(offset_milliseconds)
+
+  def write_time_micros_long(self, datum):
+    """
+    Encode python time object as long.
+    It stores the number of microseconds from midnight, 00:00:00.000000
+    """
+    offset_microseconds = (((datum.hour * 60) + datum.minute)*60 + datum.second)*1000000 + datum.microsecond
+    self._validate_time_datum(offset_microseconds, constants.TIME_MICROS)
+    self.write_long(offset_microseconds)
+
+  def _timedelta_total_seconds(self, timedelta):
+    # https://docs.python.org/2/library/datetime.html#datetime.timedelta.total_seconds
+    return (
+        timedelta.microseconds + 0.0 +
+        (timedelta.seconds + timedelta.days * 24 * 3600) * 10 ** 6) / 10 ** 6
+
+  def write_timestamp_millis_long(self, datum):
+    """
+    Encode python time object as long.
+    It stores the number of milliseconds from midnight of unix epoch, 1 January 1970.
+    """
+    timedelta = datum - datetime.datetime(1970, 1, 1, 0, 0, 0, 0)
```

Comment:
@clin7 from our offline discussion we decided to let the application developer manage the timezone information. Hence we will not support avro schema creation from python datetime object for now, application developer is responsible for appropriate conversion of datetime object to int.